Scatter chart color legend override

thangdinh opened the pull request:
### Problem
Currently the scatter chart legend does not work with color override, where we have a color field in each data point. After filtering by groups, the chart is just using the default color to set. 

### Solution
Check if there is a color override in the data point when it try to get the group color `getGroupColor`, use it. Otherwise, use the default color. 
Tests are added.

Before:
![screen shot 2016-02-23 at 8 52 08 pm](https://cloud.githubusercontent.com/assets/8838366/13275969/70111aec-da6f-11e5-8781-9efb55d0fa98.png)
 
After:
![screen shot 2016-02-23 at 8 50 53 pm](https://cloud.githubusercontent.com/assets/8838366/13275974/74f7c6e6-da6f-11e5-9a1f-b2953c5c9690.png)

@Addepar/fire @Addepar/pants @cyril-sf

jordansmith42 commented on app/templates/components/chart-component.hbs:
```diff
@@ -1,3 +1,3 @@
-  <g class="chart-viewport" transform={{transformViewport}}></g>
-</svg>
\ No newline at end of file
+<svg {{bind-attr width=outerWidth}} {{bind-attr height=outerHeight}}>
```
this should not be pushed to master, I believe this was caused by the rebase confusion

jordansmith42 commented on addon/components/scatter-chart.js:
```diff
@@ -258,11 +263,15 @@ const ScatterChartComponent = ChartComponent.extend(LegendMixin, FloatingTooltip
         group: name,
-        stroke: getGroupColor,
-        fill: displayGroups ? getGroupColor : 'transparent',
+        stroke: color,
```
Have we confirmed this doesn't break backwards compatibility?

thangdinh replied:
@jordansmith42 : It takes either a color string or a function so this change should be fine here:
https://github.com/Addepar/ember-charts/blob/master/addon/mixins/legend.js#L208-L213

thangdinh commented:
It takes either a color string or a function so this change should be fine here:
https://github.com/Addepar/ember-charts/blob/master/addon/mixins/legend.js#L208-L213

cyril-sf commented on tests/unit/scatter-test.js:
```diff
@@ -105,3 +106,18 @@ test('Test domain without domain padding', function(assert) {
 });
+
+test('Test legend color override', function(assert) {
+  var component = this.subject(scatterContent);
```
we don't seem to have a test for the case where we don't override.

metathinker replied:
I think we do, don't we? We test that the color is correct both before overriding and after overriding, and the "before overriding" case is the same as the "don't override at all" case.

Or was that part of the test added after your comment?

metathinker replied:
> Or was that part of the test added after your comment?
 
yes, looks like it, according to the "outdated diff" on the similar comment you made later:
https://github.com/Addepar/ember-charts/pull/152/files/cb4f443a829ef459b341431c2e0dd4489ca86177#r54045960

cyril-sf commented on tests/unit/scatter-test.js:
```diff
@@ -105,3 +106,18 @@ test('Test domain without domain padding', function(assert) {
+test('Test data point color override', function(assert) {
+  var component = this.subject(scatterContent);
+  var firstDataPoint = component.get('data')[0];
+  assert.equal(component.get('pointAttrs').fill(firstDataPoint,0), '#010101', 'the fill color is overriden');
```
we don't seem to have a test for the case where we don't override.

thangdinh commented:
@cyril-sf I just added tests covering default case where there is no overriding color.

jordansmith42 commented on tests/unit/scatter-test.js:
```diff
@@ -16,7 +16,7 @@ var scatterContent = {
       "group": "Energy",
       "xValue": 0.017,
-      "yValue": 0.03
+      "yValue": 0.03,
```
remove the trailing coma

jordansmith42 commented:
LGTM after comments and tests pass

billy-addepar commented:
LGTM after test passes

thangdinh commented:
The tests are failing because I changed the set function to set each individual field and it somehow showed me the tests passed before I pushed. However, `data` is not an ember object so the computed property dependency won't work. I changed it back to my initial set data. Any could help me double check it and give LGTM again? @jordansmith42 @billy-addepar

metathinker commented:
lgtm.
